src/model/validation.rs

```diff
@@ -0,0 +1,427 @@
+    /// The errors returned by the validation
+    fn errors(&self) -> Vec<&ValidationFailure>;
+}
+
+/// Default implementation for a list of concrete [`ValidationFailure`]s
+impl ValidationOutput for [ValidationFailure] {
+    fn valid(&self) -> bool {
+        self.errors().is_empty()
+    }
+    fn warnings(&self) -> Vec<&ValidationFailure> {
+        self.iter()
+            .filter(|m| m.level == ValidationFailureLevel::Warning)
+            .collect()
+    }
+    fn errors(&self) -> Vec<&ValidationFailure> {
+        self.iter()
+            .filter(|m| m.level == ValidationFailureLevel::Error)
+            .collect()
+    }
+}
+
+/// Default implementation for a list of concrete [`ValidationFailure`]s
+impl ValidationOutput for Vec<ValidationFailure> {
+    fn valid(&self) -> bool {
+        self.as_slice().valid()
+    }
+    fn warnings(&self) -> Vec<&ValidationFailure> {
+        self.iter()
+            .filter(|m| m.level == ValidationFailureLevel::Warning)
+            .collect()
+    }
+    fn errors(&self) -> Vec<&ValidationFailure> {
+        self.iter()
+            .filter(|m| m.level == ValidationFailureLevel::Error)
+            .collect()
+    }
+}
+
+/// Validate a WADM application manifest, returning a list of validation failures
+///
+/// At present this can check for:
+/// - unsupported interfaces (i.e. typos, etc)
+/// - unknown packages under known namespaces
+/// - "dangling" links (missing components)
+///
+/// Since `[ValidationFailure]` implements `ValidationOutput`, you can call `valid()` and other
+/// trait methods on it:
+///
+/// ```rust,ignore
+/// let messages = validate_manifest(some_path).await?;
+/// let valid = messages.valid();
+/// ```
+///
+/// # Arguments
+///
+/// * `path` - Path to the Manifest that will be read into memory and validated
+pub async fn validate_manifest_file(
+    path: impl AsRef<Path>,
+) -> Result<(Manifest, Vec<ValidationFailure>)> {
+    let content = tokio::fs::read_to_string(path.as_ref())
+        .await
+        .with_context(|| format!("failed to read manifest @ [{}]", path.as_ref().display()))?;
+
+    validate_manifest(&content).await.with_context(|| {
+        format!(
+            "failed to parse YAML manifest [{}]",
+            path.as_ref().display()
+        )
+    })
+}
+
+/// Validate a WADM application manifest, returning a list of validation failures
+///
+/// At present this can check for:
+/// - unsupported interfaces (i.e. typos, etc)
+/// - unknown packages under known namespaces
+/// - "dangling" links (missing components)
+///
+/// Since `[ValidationFailure]` implements `ValidationOutput`, you can call `valid()` and other
+/// trait methods on it:
+///
+/// ```rust,ignore
+/// let messages = validate_manifest(some_path).await?;
+/// let valid = messages.valid();
+/// ```
+///
+/// # Arguments
+///
+/// * `content` - YAML content  to the Manifest that will be read into memory and validated
+pub async fn validate_manifest(
```

Comment:
I think we should probably rename this one to `validate_manifest_raw` and have it take an `impl AsRef<[u8]>` and then have a `validate_manifest` function that takes a `&Manifest` parameter for the cases when a manifest has already been parsed

Reply:
Made this change, thanks! I called it `validate_manifest_bytes` though